File: api/services/task_service.py
from typing import List, Optional
from sqlmodel import Session
from models.task import Task, TaskCreate, TaskUpdate
from repositories.task_repository import TaskRepository


class TaskService:
    def __init__(self, repository: TaskRepository):
        self.repository = repository

    def create_task(self, session: Session, task: TaskCreate, user_id: str) -> Task:
        """Create a new task for the authenticated user."""
        return self.repository.create_task(session, task, user_id)

    def get_task(self, session: Session, task_id: int, user_id: str) -> Optional[Task]:
        """Get a task if it belongs to the user."""
        return self.repository.get_task(session, task_id, user_id)

    def get_all_tasks(self, session: Session, user_id: str) -> List[Task]:
        """Get all tasks for the authenticated user."""
        return self.repository.get_tasks(session, user_id)

    def update_task(self, session: Session, task_id: int, user_id: str, task_update: TaskUpdate) -> Optional[Task]:
        """Update a task if it belongs to the user."""
        task_data = task_update.model_dump(exclude_unset=True)
        return self.repository.update_task(session, task_id, user_id, task_data)

    def delete_task(self, session: Session, task_id: int, user_id: str) -> bool:
        """Delete a task if it belongs to the user."""
        return self.repository.delete_task(session, task_id, user_id)

    def complete_task(self, session: Session, task_id: int, user_id: str, completed: bool) -> Optional[Task]:
        """Toggle task completion if it belongs to the user."""
        return self.repository.update_task(session, task_id, user_id, {"completed": completed})

File: web-app/src/services/taskService.ts
import { Task, TaskCreate, TaskUpdate } from '../types/task'
import { authService } from './authService'

// API calls go directly to FastAPI backend with JWT auth
// Force relative path in production if API URL is set to localhost (common configuration error)
let apiBaseUrl = process.env.NEXT_PUBLIC_API_URL;
if (process.env.NODE_ENV === 'production' && apiBaseUrl?.includes('localhost')) {
  apiBaseUrl = '';
}
const API_BASE_URL = apiBaseUrl || (process.env.NODE_ENV === 'production' ? '' : 'http://localhost:8000');

class TaskService {
  private getHeaders(): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      ...authService.getAuthHeader(),
    };
  }

  async getAllTasks(): Promise<Task[]> {
    const response = await fetch(`${API_BASE_URL}/api/tasks`, {
      headers: this.getHeaders(),
    })
    if (!response.ok) {
      if (response.status === 401) {
        throw new Error('Please log in to view your tasks')
      }
      throw new Error('Failed to fetch tasks')
    }
    return response.json()
  }

  async getAllTasksWithFilters(search?: string, category?: string): Promise<Task[]> {
    let url = `${API_BASE_URL}/api/tasks`
    const params = new URLSearchParams()

    if (search) params.append('search', search)
    if (category) params.append('category', category)

    if (params.toString()) {
      url += `?${params.toString()}`
    }

    const response = await fetch(url, {
      headers: this.getHeaders(),
    })
    if (!response.ok) {
      throw new Error('Failed to fetch tasks')
    }
    return response.json()
  }

  async createTask(taskData: TaskCreate): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/api/tasks`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(taskData),
    })
    if (!response.ok) {
      if (response.status === 401) {
        throw new Error('Please log in to create tasks')
      }
      throw new Error('Failed to create task')
    }
    return response.json()
  }

  async updateTask(id: string, taskData: TaskUpdate): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/api/tasks/${id}`, {
      method: 'PUT',
      headers: this.getHeaders(),
      body: JSON.stringify(taskData),
    })
    if (!response.ok) {
      throw new Error('Failed to update task')
    }
    return response.json()
  }

  async completeTask(id: string, completed: boolean): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/api/tasks/${id}/complete`, {
      method: 'PATCH',
      headers: this.getHeaders(),
      body: JSON.stringify({ completed }),
    })
    if (!response.ok) {
      throw new Error('Failed to update task completion status')
    }
    return response.json()
  }

  async deleteTask(id: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/api/tasks/${id}`, {
      method: 'DELETE',
      headers: this.getHeaders(),
    })
    if (!response.ok) {
      throw new Error('Failed to delete task')
    }
  }
}

export const taskService = new TaskService()